font/src/ft/mod.rs

```diff
@@ -536,48 +583,48 @@ impl FreeTypeRasterizer {
         }
     }
 
-    fn load_face_with_glyph(&mut self, glyph: char) -> Result<FontKey, Error> {
-        let mut charset = fc::CharSet::new();
-        charset.add(glyph);
-        let mut pattern = fc::Pattern::new();
-        pattern.add_charset(&charset);
-        pattern.add_pixelsize(self.pixel_size as f64);
-
-        let config = fc::Config::get_current();
-        match fc::font_match(config, &mut pattern) {
-            Some(pattern) => {
-                if let (Some(path), Some(_)) = (pattern.file(0), pattern.index().nth(0)) {
-                    match self.keys.get(&path) {
-                        // We've previously loaded this font, so don't
-                        // load it again.
-                        Some(&key) => {
-                            debug!("Hit for font {:?}; no need to load", path);
-                            // Update fixup factor
-                            self.faces.get_mut(&key).unwrap().pixelsize_fixup_factor =
-                                pattern.pixelsizefixupfactor().next();
-                            Ok(key)
-                        },
-
-                        None => {
-                            debug!("Miss for font {:?}; loading now", path);
-                            // Safe to unwrap the option since we've already checked for the path
-                            // and index above.
-                            let key = self.face_from_pattern(&pattern)?.unwrap();
-                            Ok(key)
-                        },
+    fn load_face_with_glyph(&mut self, glyph: GlyphKey) -> Result<FontKey, Error> {
+        let ch = glyph.c;
+        let fallback_list = self.fallback_lists.get(&glyph.font_key).unwrap();
+        for font_pattern in fallback_list {
+            if let Some(path) = font_pattern.file(0) {
+                match self.keys.get(&path) {
+                    Some(&key) => {
+                        if let Some(face) =  self.faces.get(&key) {
+                            let index = face.ft_face.get_char_index(ch as usize);
+                            // We found something in a current face, so let's use it
+                            if index != 0 {
+                                return Ok(key);
+                            }
+                        }
+                    },
+                    None => {
+                        // This code will be hit, only if you don't have glyph at all or didn't
+                        // loaded font, so not that hot for a normal scenario and all costly parts
+                        // are inside the if
```

Comment:
If I understand this correctly, this code will load the font once, if it hasn't been loaded yet?

Why don't we do this when the fallback list is loaded initially for all fonts in that list? Is it to minimize the number of fonts we have to load, so on a system with 150 fallback fonts we don't load 150 fonts but instead just load the first few (assuming those cover most characters)?

Reply:
Yes, it'll load once when you're facing a font that hasn't been loaded in your fallback list *and* the char you're requesting is presented in this font. This code is fast and there's no desire to build all faces for all fonts during load_font. I don't want to load 150 faces for fonts that we wont use. Like I can work only with ASCII most of time, why should I load all the things, it's just pointless.

Reply:
Yeah, makes a lot of sense. Though it is a bit messy. Might be good to at least put that in another function with a descriptive name, so the initialization of a font isn't just sitting in some random match statement somewhere.

Reply:
There are still some things I want to try on this branch "path", so I'll leave it here as it is for now, and come back later.